hoot-core/src/main/cpp/hoot/core/conflate/network/EdgeMatchSetFinder.cpp

```diff
@@ -428,12 +373,73 @@ void EdgeMatchSetFinder::_prependMatch(EdgeMatchPtr em, ConstNetworkEdgePtr e1,
 {
-  bool result = false;
-  double score = _scoreMatch(em);
+  LOG_TRACE("Recording match: " << em);
+
+  const double score = _scoreMatch(em);
   LOG_VART(score);
   if (score > 0)
   {
-    LOG_TRACE("Recording match: " << em);
+    // We want to avoid adding matches that are very similar to ones we've already found and have
+    // lower score than they do (see EdgeMatch::isVerySimilarTo).  Otherwise we'll have to remove
+    // them before calculating match relationships, which can be very expensive.
+
+    QMap<QString, QString> similarityValuesToIterate;
+    similarityValuesToIterate[EDGE_MATCH_SIMILAR_KEY] = em->getSimilarityString();
+    similarityValuesToIterate[EDGE_MATCH_SIMILAR_SECOND_REVERSED_KEY] =
+      em->getSecondReversedSimilarityString();
+    similarityValuesToIterate[EDGE_MATCH_SIMILAR_FIRST_REVERSED_KEY] =
+      em->getFirstReversedSimilarityString();
+    EdgeMatchScore existingSimilarMatch;
+    QMap<QString, QString>::iterator similarityValsItr;
+    // This will iterate over the similarity indexes in the order we want.
+    for (similarityValsItr = similarityValuesToIterate.begin();
+         similarityValsItr != similarityValuesToIterate.end(); ++similarityValsItr)
+    {
+      existingSimilarMatch =
+        _edgeMatchSimilarities[similarityValsItr.key()][similarityValsItr.value()];
+      if (existingSimilarMatch.score != -1.0)
+      {
+        break;
+      }
+    }
+
+    // An EdgeMatchScore returned with a score == -1.0 means that no similar match was found.
+    if (existingSimilarMatch.score != -1.0)
+    {
+      // If we already have an edge with a higher score that is very similar to this edge, then
+      // don't add it.
+      if (existingSimilarMatch.score >= score)
+      {
+        LOG_TRACE(
+          "Found edge match: " << em << " similar to existing match: " <<
+          existingSimilarMatch.match << " and with lower score: " << score <<
+          ".  Skipping match...");
+        _numSimilarEdgeMatches++;
+        // Returning success here is slightly counterintuitive but is the correct thing to do for
+        // the match alg.
+        return true;
+      }
+      // Otherwise, remove the existing match, since the new one has a higher score.
+      else
+      {
+        LOG_TRACE(
+          "Removing similar edge match: " << existingSimilarMatch.match << " with lower score: " <<
+          score << " than found edge match: " << em << "...");;
+        _matchSet->removeEdgeMatch(existingSimilarMatch.match);
+      }
+    }
+
+    // similarity index our new match (overwrites the index of any existing matches)
+    EdgeMatchScore newMatch;
+    newMatch.match = em;
+    newMatch.score = score;
+    for (similarityValsItr = similarityValuesToIterate.begin();
+         similarityValsItr != similarityValuesToIterate.end(); ++similarityValsItr)
+    {
+      _edgeMatchSimilarities[similarityValsItr.key()][similarityValsItr.value()] = newMatch;
+    }
```

Comment:
This is great, keeping track of the similarity scores and not adding lower scores and updating with higher scores.